File: android/DJMediWallet/app/src/main/java/com/djmediwallet/models/fhir/Patient.kt
package com.djmediwallet.models.fhir

/**
 * FHIR Resource wrapper
 */
data class FHIRResource(
    val resourceType: String,
    val id: String? = null,
    val data: Map<String, Any>? = null
)

/**
 * FHIR Patient resource
 */
data class Patient(
    val resourceType: String = "Patient",
    val id: String? = null,
    val identifier: List<Identifier>? = null,
    val active: Boolean? = true,
    val name: List<HumanName>? = null,
    val telecom: List<ContactPoint>? = null,
    val gender: String? = null,
    val birthDate: String? = null,
    val address: List<Address>? = null,
    val contact: List<PatientContact>? = null
)

/**
 * FHIR Identifier
 */
data class Identifier(
    val system: String? = null,
    val value: String? = null,
    val use: String? = null,
    val type: CodeableConcept? = null
)

/**
 * FHIR HumanName
 */
data class HumanName(
    val use: String? = null,
    val family: String? = null,
    val given: List<String>? = null,
    val prefix: List<String>? = null,
    val suffix: List<String>? = null
) {
    val fullName: String
        get() {
            val parts = mutableListOf<String>()
            prefix?.let { parts.addAll(it) }
            given?.let { parts.addAll(it) }
            family?.let { parts.add(it) }
            suffix?.let { parts.addAll(it) }
            return parts.joinToString(" ")
        }
}

/**
 * FHIR ContactPoint
 */
data class ContactPoint(
    val system: String? = null,
    val value: String? = null,
    val use: String? = null
)

/**
 * FHIR Address
 */
data class Address(
    val use: String? = null,
    val type: String? = null,
    val line: List<String>? = null,
    val city: String? = null,
    val state: String? = null,
    val postalCode: String? = null,
    val country: String? = null
) {
    val fullAddress: String
        get() {
            val parts = mutableListOf<String>()
            line?.let { parts.addAll(it) }
            city?.let { parts.add(it) }
            state?.let { parts.add(it) }
            postalCode?.let { parts.add(it) }
            country?.let { parts.add(it) }
            return parts.joinToString(", ")
        }
}

/**
 * FHIR Patient Contact
 */
data class PatientContact(
    val relationship: List<CodeableConcept>? = null,
    val name: HumanName? = null,
    val telecom: List<ContactPoint>? = null,
    val address: Address? = null
)

/**
 * FHIR CodeableConcept
 */
data class CodeableConcept(
    val coding: List<Coding>? = null,
    val text: String? = null
) {
    val display: String?
        get() = text ?: coding?.firstOrNull()?.display
}

/**
 * FHIR Coding
 */
data class Coding(
    val system: String? = null,
    val code: String? = null,
    val display: String? = null,
    val version: String? = null
) {
    val isSNOMED: Boolean
        get() = system == "http://snomed.info/sct"
}

File: ios/DJMediWallet/DJMediWalletXCode/DJMediWallet25/DJMediWallet25/Models/FHIR/Patient.swift
//
//  Patient.swift
//  DJMediWallet
//
//  FHIR Patient resource model
//

import Foundation

/// FHIR Patient resource
public struct Patient: Codable {
    public let resourceType: String
    public let id: String?
    public let identifier: [Identifier]?
    public let active: Bool?
    public let name: [HumanName]?
    public let telecom: [ContactPoint]?
    public let gender: String?
    public let birthDate: String?
    public let address: [Address]?
    public let contact: [PatientContact]?
    
    public init(
        id: String? = nil,
        identifier: [Identifier]? = nil,
        active: Bool? = true,
        name: [HumanName]? = nil,
        telecom: [ContactPoint]? = nil,
        gender: String? = nil,
        birthDate: String? = nil,
        address: [Address]? = nil,
        contact: [PatientContact]? = nil
    ) {
        self.resourceType = "Patient"
        self.id = id
        self.identifier = identifier
        self.active = active
        self.name = name
        self.telecom = telecom
        self.gender = gender
        self.birthDate = birthDate
        self.address = address
        self.contact = contact
    }
}

/// FHIR Identifier
public struct Identifier: Codable {
    public let system: String?
    public let value: String?
    public let use: String?
    public let type: CodeableConcept?
    
    public init(system: String?, value: String?, use: String? = nil, type: CodeableConcept? = nil) {
        self.system = system
        self.value = value
        self.use = use
        self.type = type
    }
}

/// FHIR HumanName
public struct HumanName: Codable {
    public let use: String?
    public let family: String?
    public let given: [String]?
    public let prefix: [String]?
    public let suffix: [String]?
    
    public init(use: String? = nil, family: String?, given: [String]?, prefix: [String]? = nil, suffix: [String]? = nil) {
        self.use = use
        self.family = family
        self.given = given
        self.prefix = prefix
        self.suffix = suffix
    }
    
    public var fullName: String {
        var parts: [String] = []
        if let prefix = prefix {
            parts.append(contentsOf: prefix)
        }
        if let given = given {
            parts.append(contentsOf: given)
        }
        if let family = family {
            parts.append(family)
        }
        if let suffix = suffix {
            parts.append(contentsOf: suffix)
        }
        return parts.joined(separator: " ")
    }
}

/// FHIR ContactPoint
public struct ContactPoint: Codable {
    public let system: String?
    public let value: String?
    public let use: String?
    
    public init(system: String?, value: String?, use: String? = nil) {
        self.system = system
        self.value = value
        self.use = use
    }
}

/// FHIR Address
public struct Address: Codable {
    public let use: String?
    public let type: String?
    public let line: [String]?
    public let city: String?
    public let state: String?
    public let postalCode: String?
    public let country: String?
    
    public init(
        use: String? = nil,
        type: String? = nil,
        line: [String]? = nil,
        city: String? = nil,
        state: String? = nil,
        postalCode: String? = nil,
        country: String? = nil
    ) {
        self.use = use
        self.type = type
        self.line = line
        self.city = city
        self.state = state
        self.postalCode = postalCode
        self.country = country
    }
    
    public var fullAddress: String {
        var parts: [String] = []
        if let line = line {
            parts.append(contentsOf: line)
        }
        if let city = city {
            parts.append(city)
        }
        if let state = state {
            parts.append(state)
        }
        if let postalCode = postalCode {
            parts.append(postalCode)
        }
        if let country = country {
            parts.append(country)
        }
        return parts.joined(separator: ", ")
    }
}

/// FHIR Patient Contact
public struct PatientContact: Codable {
    public let relationship: [CodeableConcept]?
    public let name: HumanName?
    public let telecom: [ContactPoint]?
    public let address: Address?
    
    public init(
        relationship: [CodeableConcept]? = nil,
        name: HumanName? = nil,
        telecom: [ContactPoint]? = nil,
        address: Address? = nil
    ) {
        self.relationship = relationship
        self.name = name
        self.telecom = telecom
        self.address = address
    }
}

/// FHIR CodeableConcept
public struct CodeableConcept: Codable {
    public let coding: [Coding]?
    public let text: String?
    
    public init(coding: [Coding]?, text: String? = nil) {
        self.coding = coding
        self.text = text
    }
    
    public var display: String? {
        return text ?? coding?.first?.display
    }
}

/// FHIR Coding
public struct Coding: Codable {
    public let system: String?
    public let code: String?
    public let display: String?
    public let version: String?
    
    public init(system: String?, code: String?, display: String? = nil, version: String? = nil) {
        self.system = system
        self.code = code
        self.display = display
        self.version = version
    }
    
    public var isSNOMED: Bool {
        return system == "http://snomed.info/sct"
    }
}
